src/Microsoft.Diagnostics.Monitoring.WebApi/Stacks/EventStacksPipeline.cs

```diff
@@ -0,0 +1,153 @@
+
+        public EventStacksPipeline(DiagnosticsClient client, EventStacksPipelineSettings settings)
+            : base(client, settings)
+        {
+        }
+
+        protected override MonitoringSourceConfiguration CreateConfiguration()
+        {
+            return new EventPipeProviderSourceConfiguration(requestRundown: false, bufferSizeInMB: 256, new[]
+            {
+                new EventPipeProvider(StackEvents.Provider, System.Diagnostics.Tracing.EventLevel.LogAlways)
+            });
+        }
+
+        protected override async Task OnEventSourceAvailable(EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token)
+        {
+            eventSource.Dynamic.AddCallbackForProviderEvents((string provider, string _) => provider == StackEvents.Provider ?
+            EventFilterResponse.AcceptEvent : EventFilterResponse.RejectEvent, Callback);
+
+            using EventTaskSource<Action> sourceComplete = new EventTaskSource<Action>(
+                taskComplete => taskComplete,
+                addHandler => eventSource.Completed += addHandler,
+                removeHandler => eventSource.Completed -= removeHandler,
+                token);
+
+            // This is the same issue as GCDumps. We don't always get events back in realtime, so we have to stop the session and then process the events.
+            Task eventsTimeoutTask = Task.Delay(Settings.Timeout, token);
+            Task completedTask = await Task.WhenAny(_stackResult.Task, eventsTimeoutTask);
+
+            await completedTask;
+
+            await stopSessionAsync();
+            await sourceComplete.Task;
+
+            if (_stackResult.Task.Status != TaskStatus.RanToCompletion)
+            {
+                throw new InvalidOperationException("Unable to process stack in timely manner.");
+            }
+        }
+
+        public Task<StackResult> Result => _stackResult.Task;
+
+        private void Callback(TraceEvent action)
+        {
+            if (action.ProviderName != StackEvents.Provider)
+            {
+                return;
+            }
+
+            //We do not have a manifest for our events, but we also lookup data by id instead of string.
+            if (action.ID == StackEvents.Callstack)
+            {
+                var stack = new Stack
+                {
+                    ThreadId = action.GetPayload<uint>(StackEvents.CallstackPayloads.ThreadId)
+                };
+                ulong[] functionIds = action.GetPayload<ulong[]>(StackEvents.CallstackPayloads.FunctionIds);
+                ulong[] offsets = action.GetPayload<ulong[]>(StackEvents.CallstackPayloads.IpOffsets);
+
+                _result.Stacks.Add(stack);
+
+                if (functionIds != null && offsets != null)
+                {
+                    for (int i = 0; i < functionIds.Length; i++)
+                    {
+                        stack.Frames.Add(new StackFrame { FunctionId = functionIds[i], Offset = offsets[i] });
```

Comment:
I think Justin already mentioned this on his comment above, but is there any chance that `functionIds` and `offsets` are out of sync, in which case we could potentially overrun the end of `offsets`?